src/IK/src/InverseKinematics.cpp

```diff
@@ -322,43 +322,107 @@ bool HumanIK::updateFloorContactTasks(const std::unordered_map<int, Eigen::Matri
     return true;
 }
 
-bool HumanIK::TPoseCalibrationNode(const int node, const manif::SO3d& I_R_IMU)
+bool HumanIK::calibrateWorldYaw(std::unordered_map<int, nodeData> nodeStruct)
 {
-    m_tPose = true;
-    // check if the node number is valid
-    if ((m_OrientationTasks.find(node) == m_OrientationTasks.end()) && (m_GravityTasks.find(node) == m_GravityTasks.end()))
+    // reset the robot state
+    Eigen::VectorXd jointPositions;
+    Eigen::VectorXd jointVelocities;
+    jointPositions.resize(this->getDoFsNumber());
+    jointPositions.setZero();
+    jointVelocities.resize(this->getDoFsNumber());
+    jointVelocities.setZero();
+    Eigen::Matrix4d basePose;
+    basePose.setIdentity();
+    Eigen::VectorXd baseVelocity;
+    baseVelocity.resize(6);
+    baseVelocity.setZero();
+    m_kinDyn->setRobotState(basePose, jointPositions, baseVelocity, jointVelocities, m_gravity);
+    // Update the orientation and gravity tasks
+    for (const auto& [node, data] : nodeStruct)
     {
-        BiomechanicalAnalysis::log()->error("[HumanIK::TPoseCalibrationNode] Invalid node number.");
-        return false;
-    }
+        // check if the node number is valid
+        if (m_OrientationTasks.find(node) == m_OrientationTasks.end() && m_GravityTasks.find(node) == m_GravityTasks.end())
+        {
+            BiomechanicalAnalysis::log()->error("[HumanIK::calibrateWorldYaw] Invalid node number.");
+            return false;
+        }
 
-    // compute the rotation matrix from the world to the world of the IMU as:
-    // W_R_WIMU = R_calib * (WIMU_R_IMU * IMU_R_link)^{T}
-    // where R_calib is assumed to be the identity
-    // The if condition checks whether the task is m_OrientationTasks or m_GravityTasks
-    if (m_OrientationTasks.find(node) != m_OrientationTasks.end())
-    {
-        m_OrientationTasks[node].calibrationMatrix = calib_W_R_link * (I_R_IMU * m_OrientationTasks[node].IMU_R_link).inverse();
-    } else
-    {
-        m_GravityTasks[node].calibrationMatrix = calib_W_R_link * (I_R_IMU * m_GravityTasks[node].IMU_R_link).inverse();
+        iDynTree::Rotation rpyOffset;
+        if (m_OrientationTasks.find(node) != m_OrientationTasks.end())
+        {
+            // compute the offset between the world and the IMU world as:
+            // W_R_WIMU = W_R_link * (WIMU_R_IMU * IMU_R_link)^{T}
+            iDynTree::toEigen(rpyOffset) = iDynTree::toEigen(m_kinDyn->getWorldTransform(m_OrientationTasks[node].frameName).getRotation())
+                                           * ((data.I_R_IMU * m_OrientationTasks[node].IMU_R_link).inverse()).rotation();
+            // set the calibration matrix of the orientation task to the offset in yaw
+            m_OrientationTasks[node].calibrationMatrix = manif::SO3d(0, 0, rpyOffset.asRPY()(2));
```

Comment:
Should we write somewhere that we assume the gravity to be on the Z axis?

Reply:
done in https://github.com/ami-iit/biomechanical-analysis-framework/pull/45/commits/d989bc71410a7b49eb2d2db0afe727960cb77973